lib/Optimizer/GraphOptimizer/GraphOptimizer.cpp

```diff
@@ -2542,6 +2542,52 @@ bool FoldMatMulAddIntoFullyConnected::run(Function *F,
   return changed;
 }
 
+/// Convert MatMul into FullyConnected with null bias. This pass is used if
+/// we have an optimized implementation for FullyConnected but NOT for MatMul.
+/// Make sure you run this pass after the FoldMatMulAddIntoFullyConnected pass
+/// otherwise a MatMul followed by Add will be converted into a FullyConnected
+/// followed by Add and NOT a single FullyConnected instance.
+bool ConvertMatMulToFullyConnected::run(Function *F,
+                                        const CompilationContext &cctx) {
+  bool changed = false;
+  for (auto &node : F->getNodes()) {
+    auto *matMulNode = dyn_cast<MatMulNode>(&node);
+    if (!matMulNode) {
+      continue;
+    }
+
+    // Create null bias.
+    Constant *bias = nullptr;
+    std::vector<dim_t> biasDims = {matMulNode->getResult().dims().back()};
+    std::string biasName = matMulNode->getName().str() + "bias";
+    if (matMulNode->getResult().getType()->isQuantizedType()) {
+      // Create null bias with offset 0 and a scale equal to the product
+      // between LHS scale and RHS scale.
+      float biasScale = matMulNode->getLHS().getType()->getScale() *
+                        matMulNode->getRHS().getType()->getScale();
+      int32_t biasOffset = 0;
+      ElemKind biasPrec = cctx.precisionConfig.quantConfig.precisionBias;
+      bias = F->getParent()->createConstant(biasPrec, biasDims, biasScale,
+                                            biasOffset, biasName);
+      bias->getPayloadMutable().zero();
+    } else {
+      // Create null FLOAT bias.
+      bias =
+          F->getParent()->createConstant(ElemKind::FloatTy, biasDims, biasName);
+      bias->getPayloadMutable().zero();
```

Comment:
Not a huge deal, but you could have used Splat nodes here instead of Constants that have value=0. Then size of constants for the model is a bit smaller (though I think Splats essentially gets just baked in as constants to the code for LLVM-based backends). They also could get materialized to Constants later on, depending on `cctx.optimizationOpts.materializeSplatsUsedBySet`. Can still land as is if you really prefer Constants.

Reply:
Yeah it makes sense but we currently use this pass for our backends expecting some null constants to be produc so I would leave this as-is:
- the bias operands are usually much smaller than the (say) weights for MatMul so using Constants instead of Splat will not increase significantly the model size
- this pass is currently optional (backends opt-in to use this pass) so we could make it more flexible later on when we will need higher flexibility (e.g. we could modify the pass itself to create Splat/Constant based on the value of the `cctx.optimizationOpts.materializeSplatsUsedBySet` parameters)

Reply:
Yeah it makes sense but we currently use this pass for our backends expecting some null constants to be produc so I would leave this as-is:
- the bias operands are usually much smaller than the (say) weights for MatMul so using Constants instead of Splat will not increase significantly the model size
- this pass is currently optional (backends opt-in to use this pass) so we could make it more flexible later on when we will need higher flexibility (e.g. we could modify the pass itself to create Splat/Constant based on the value of the `cctx.optimizationOpts.materializeSplatsUsedBySet` parameters)